File: apps/ios/PagApp/PagApp/Services/RewardService.swift
import Foundation
import Combine
import FirebaseFunctions

@MainActor
public class RewardService: ObservableObject {
    public static let shared = RewardService()
    
    @Published public private(set) var rewardBalance: Int = 0
    @Published public private(set) var profileScore: Int = 0
    @Published public private(set) var rewards: [PAGRewardLedgerEntry] = []
    @Published public private(set) var vouchers: [PAGVoucher] = []
    @Published public private(set) var scoreLedgers: [PAGScoreLedgerEntry] = []
    @Published public private(set) var isLoading: Bool = false
    @Published public private(set) var errorMessage: String? = nil
    
    private init() {}
    
    public func fetchUserRewards() async {
        if rewards.isEmpty && vouchers.isEmpty && scoreLedgers.isEmpty {
            self.isLoading = true
        }
        self.errorMessage = nil
        
        do {
            // 1. Try High-Speed REST API (~10ms)
            if let response = try? await PAGApiClient.shared.get(endpoint: "/wallet"),
               let success = response["success"] as? Bool, success,
               let dataDict = response["data"] as? [String: Any] {
                
                self.rewardBalance = Int(Double(dataDict["rewardBalance"] as? String ?? "0") ?? 0)
                self.profileScore = dataDict["profileScore"] as? Int ?? 0
                
                var parsedRewards: [PAGRewardLedgerEntry] = []
                if let rawRewards = dataDict["rewardHistory"] as? [[String: Any]] {
                    for r in rawRewards {
                        let entry = PAGRewardLedgerEntry(
                            id: r["id"] as? String ?? UUID().uuidString,
                            surveyId: r["surveyId"] as? String ?? "",
                            type: r["rewardType"] as? String ?? "MONEY",
                            amount: Int(Double(r["amount"] as? String ?? "0") ?? 0),
                            reason: "Anket Ödülü",
                            createdAt: r["createdAt"] as? String ?? ""
                        )
                        parsedRewards.append(entry)
                    }
                }
                self.rewards = parsedRewards
                
                var parsedVouchers: [PAGVoucher] = []
                if let rawVouchers = dataDict["vouchers"] as? [[String: Any]] {
                    for v in rawVouchers {
                        let voucher = PAGVoucher(
                            voucherId: v["id"] as? String ?? UUID().uuidString,
                            poolId: v["surveyId"] as? String ?? "",
                            title: v["poolName"] as? String ?? "Hediye Çeki",
                            code: v["code"] as? String ?? "",
                            valueAmount: Int(Double(v["amount"] as? String ?? "0") ?? 0),
                            status: v["status"] as? String ?? "ASSIGNED",
                            assignedAt: v["assignedAt"] as? String ?? "",
                            expiresAt: nil
                        )
                        parsedVouchers.append(voucher)
                    }
                }
                self.vouchers = parsedVouchers
                
                var parsedScoreLedgers: [PAGScoreLedgerEntry] = []
                if let rawScores = dataDict["scoreHistory"] as? [[String: Any]] {
                    for s in rawScores {
                        let entry = PAGScoreLedgerEntry(
                            id: s["id"] as? String ?? UUID().uuidString,
                            userId: s["userId"] as? String ?? "",
                            sourceType: s["sourceType"] as? String ?? "SURVEY",
                            sourceId: s["sourceId"] as? String ?? "",
                            amount: s["scoreDelta"] as? Int ?? 0,
                            reason: "Profil Puanı",
                            createdAt: s["createdAt"] as? String ?? ""
                        )
                        parsedScoreLedgers.append(entry)
                    }
                }
                self.scoreLedgers = parsedScoreLedgers
                self.isLoading = false
                return
            }

            // 2. Fallback to Firebase Callable
            let result = try await Functions.functions().httpsCallable("getUserRewards").call()
            guard let dict = result.data as? [String: Any],
                  let success = dict["success"] as? Bool, success,
                  let dataDict = dict["data"] as? [String: Any] else {
                throw NSError(domain: "RewardService", code: 500, userInfo: [NSLocalizedDescriptionKey: "Ödüller yüklenemedi."])
            }
            
            self.rewardBalance = dataDict["rewardBalance"] as? Int ?? 0
            self.profileScore = dataDict["profileScore"] as? Int ?? 0
            
            // 1. Monetary Rewards
            var parsedRewards: [PAGRewardLedgerEntry] = []
            if let rawRewards = dataDict["rewards"] as? [[String: Any]] {
                for r in rawRewards {
                    let entry = PAGRewardLedgerEntry(
                        id: r["id"] as? String ?? UUID().uuidString,
                        surveyId: r["surveyId"] as? String ?? "",
                        type: r["type"] as? String ?? "MONEY",
                        amount: r["amount"] as? Int ?? 0,
                        reason: r["reason"] as? String ?? "Anket Ödülü",
                        createdAt: r["createdAt"] as? String ?? ""
                    )
                    parsedRewards.append(entry)
                }
            }
            self.rewards = parsedRewards
            
            // 2. Vouchers
            var parsedVouchers: [PAGVoucher] = []
            if let rawVouchers = dataDict["vouchers"] as? [[String: Any]] {
                for v in rawVouchers {
                    let voucher = PAGVoucher(
                        voucherId: v["voucherId"] as? String ?? UUID().uuidString,
                        poolId: v["poolId"] as? String ?? "",
                        title: v["title"] as? String ?? "Hediye Çeki",
                        code: v["code"] as? String ?? "",
                        valueAmount: v["valueAmount"] as? Int ?? 0,
                        status: v["status"] as? String ?? "ASSIGNED",
                        assignedAt: v["assignedAt"] as? String ?? "",
                        expiresAt: v["expiresAt"] as? String
                    )
                    parsedVouchers.append(voucher)
                }
            }
            self.vouchers = parsedVouchers
            
            // 3. Profile Score History
            var parsedScoreLedgers: [PAGScoreLedgerEntry] = []
            if let rawScores = dataDict["scoreLedgers"] as? [[String: Any]] {
                for s in rawScores {
                    let entry = PAGScoreLedgerEntry(
                        id: s["id"] as? String ?? UUID().uuidString,
                        userId: s["userId"] as? String ?? "",
                        sourceType: s["sourceType"] as? String ?? "SURVEY",
                        sourceId: s["sourceId"] as? String ?? "",
                        amount: s["amount"] as? Int ?? 0,
                        reason: s["reason"] as? String ?? "Profil Puanı",
                        createdAt: s["createdAt"] as? String ?? ""
                    )
                    parsedScoreLedgers.append(entry)
                }
            }
            self.scoreLedgers = parsedScoreLedgers
            self.isLoading = false
        } catch {
            print("[RewardService] Fetch rewards error: \(error.localizedDescription)")
            self.errorMessage = "Ödül verileri yüklenirken bir sorun oluştu."
            self.isLoading = false
        }
    }
}

File: apps/android/app/src/main/java/com/alafteknoloji/pagapp/services/RewardService.kt
package com.alafteknoloji.pagapp.services

import com.alafteknoloji.pagapp.models.PAGRewardLedgerEntry
import com.alafteknoloji.pagapp.models.PAGVoucher
import com.google.firebase.functions.FirebaseFunctions
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.tasks.await

data class PAGScoreLedgerEntry(
    val id: String = "",
    val sourceType: String = "BASIC_PROFILE",
    val amount: Int = 10,
    val reason: String = "Profil Puanı Ödülü",
    val createdAt: String = ""
)

class RewardService {
    private val functions: FirebaseFunctions = FirebaseFunctions.getInstance()

    private val _rewardBalance = MutableStateFlow(0)
    val rewardBalance: StateFlow<Int> = _rewardBalance.asStateFlow()

    private val _profileScore = MutableStateFlow(0)
    val profileScore: StateFlow<Int> = _profileScore.asStateFlow()

    private val _rewardLedgers = MutableStateFlow<List<PAGRewardLedgerEntry>>(emptyList())
    val rewardLedgers: StateFlow<List<PAGRewardLedgerEntry>> = _rewardLedgers.asStateFlow()

    private val _vouchers = MutableStateFlow<List<PAGVoucher>>(emptyList())
    val vouchers: StateFlow<List<PAGVoucher>> = _vouchers.asStateFlow()

    private val _scoreLedgers = MutableStateFlow<List<PAGScoreLedgerEntry>>(emptyList())
    val scoreLedgers: StateFlow<List<PAGScoreLedgerEntry>> = _scoreLedgers.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    suspend fun fetchUserRewards() {
        _isLoading.value = true
        try {
            val apiRes = PAGApiClient.get("/wallet")
            if (apiRes != null && apiRes.optBoolean("success")) {
                val dataDict = apiRes.optJSONObject("data")
                if (dataDict != null) {
                    _rewardBalance.value = dataDict.optString("rewardBalance").toDoubleOrNull()?.toInt() ?: 0
                    _profileScore.value = dataDict.optInt("profileScore", 0)

                    val rawLedgers = dataDict.optJSONArray("rewardHistory")
                    val parsedLedgers = mutableListOf<PAGRewardLedgerEntry>()
                    if (rawLedgers != null) {
                        for (i in 0 until rawLedgers.length()) {
                            val l = rawLedgers.getJSONObject(i)
                            parsedLedgers.add(
                                PAGRewardLedgerEntry(
                                    id = l.optString("id"),
                                    surveyId = l.optString("surveyId"),
                                    type = l.optString("rewardType", "MONEY"),
                                    amount = l.optString("amount").toDoubleOrNull()?.toInt() ?: 0,
                                    reason = "Anket Ödülü",
                                    createdAt = l.optString("createdAt")
                                )
                            )
                        }
                    }
                    _rewardLedgers.value = parsedLedgers

                    val rawVouchers = dataDict.optJSONArray("vouchers")
                    val parsedVouchers = mutableListOf<PAGVoucher>()
                    if (rawVouchers != null) {
                        for (i in 0 until rawVouchers.length()) {
                            val v = rawVouchers.getJSONObject(i)
                            parsedVouchers.add(
                                PAGVoucher(
                                    voucherId = v.optString("id"),
                                    poolId = v.optString("surveyId"),
                                    title = v.optString("poolName", "Hediye Çeki"),
                                    code = v.optString("code"),
                                    valueAmount = v.optString("amount").toDoubleOrNull()?.toInt() ?: 0,
                                    status = v.optString("status", "ASSIGNED"),
                                    assignedAt = v.optString("assignedAt"),
                                    expiresAt = null
                                )
                            )
                        }
                    }
                    _vouchers.value = parsedVouchers

                    val rawScores = dataDict.optJSONArray("scoreHistory")
                    val parsedScores = mutableListOf<PAGScoreLedgerEntry>()
                    if (rawScores != null) {
                        for (i in 0 until rawScores.length()) {
                            val s = rawScores.getJSONObject(i)
                            parsedScores.add(
                                PAGScoreLedgerEntry(
                                    id = s.optString("id"),
                                    sourceType = s.optString("sourceType", "SURVEY"),
                                    amount = s.optInt("scoreDelta", 0),
                                    reason = "Profil Puanı",
                                    createdAt = s.optString("createdAt")
                                )
                            )
                        }
                    }
                    _scoreLedgers.value = parsedScores
                }
            } else {
                _rewardBalance.value = 0
                _profileScore.value = 0
                _rewardLedgers.value = emptyList()
                _vouchers.value = emptyList()
                _scoreLedgers.value = emptyList()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            _rewardBalance.value = 0
            _profileScore.value = 0
            _rewardLedgers.value = emptyList()
            _vouchers.value = emptyList()
            _scoreLedgers.value = emptyList()
        } finally {
            _isLoading.value = false
        }
    }
}
